arm/compute/client.go

```diff
@@ -0,0 +1,2031 @@
+//
+// resourceGroupName is the name of the resource group. name is parameters
+// supplied to the Create Availability Set operation. parameters is
+// parameters supplied to the Create Availability Set operation.
+func (client *AvailabilitySetsClient) CreateOrUpdate(resourceGroupName string, name string, parameters AvailabilitySet) (result AvailabilitySet, err error) {
+
+	req, err := client.NewCreateOrUpdateRequest(resourceGroupName, name, parameters)
+	if err != nil {
+		return result, fmt.Errorf("compute: Failure preparing AvailabilitySetsClient.CreateOrUpdate request (%v)", err)
+	}
+
+	req, err = autorest.Prepare(
+		req,
+		client.WithAuthorization(),
+		client.WithInspection())
+	if err != nil {
+		return result, fmt.Errorf("compute: Failure sending AvailabilitySetsClient.CreateOrUpdate request (%v)", err)
+	}
+
+	resp, err := autorest.SendWithSender(client, req)
+	if err != nil {
+		return result, fmt.Errorf("compute: Failure sending AvailabilitySetsClient.CreateOrUpdate request (%v)", err)
+	}
+
+	result = AvailabilitySet{}
+	err = autorest.Respond(
+		resp,
+		client.ByInspecting(),
+		autorest.WithErrorUnlessOK(),
+		autorest.ByUnmarshallingJSON(&result),
+		autorest.ByClosing())
+
+	return result, err
+}
+
+func (client *AvailabilitySetsClient) NewCreateOrUpdateRequest(resourceGroupName string, name string, parameters AvailabilitySet) (*http.Request, error) {
+
+	pathParameters := map[string]interface{}{
+		"name":              url.QueryEscape(name),
+		"resourceGroupName": url.QueryEscape(resourceGroupName),
+		"subscriptionId":    url.QueryEscape(client.SubscriptionId),
+	}
+
+	queryParameters := map[string]interface{}{
+		"api-version": ApiVersion,
+	}
+
+	return autorest.DecoratePreparer(
+		client.CreateOrUpdateRequestPreparer(),
+		autorest.WithJSON(parameters),
+		autorest.WithPathParameters(pathParameters),
+		autorest.WithQueryParameters(queryParameters)).Prepare(&http.Request{})
+}
+
+func (client *AvailabilitySetsClient) CreateOrUpdateRequestPreparer() autorest.Preparer {
+	return autorest.CreatePreparer(
+		autorest.AsJSON(),
+		autorest.AsPut(),
+		autorest.WithBaseURL(client.BaseUri),
+		autorest.WithPath("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/availabilitySets/{name}"))
+}
+
+////////////////////////////////////////////////////////////////////////////////
+//
+// VirtualMachineImages Client
+//
+////////////////////////////////////////////////////////////////////////////////
+type VirtualMachineImagesClient struct {
+	ComputeManagementClient
+}
+
+func NewVirtualMachineImagesClient(subscriptionId string) *VirtualMachineImagesClient {
+	return NewVirtualMachineImagesClientWithBaseUri(DefaultBaseUri, subscriptionId)
+}
+
+func NewVirtualMachineImagesClientWithBaseUri(baseUri string, subscriptionId string) *VirtualMachineImagesClient {
+	return &VirtualMachineImagesClient{*NewWithBaseUri(baseUri, subscriptionId)}
```

Comment:
Ok we discussed this vaguely before. I don't think returning `*FooClient` has any advantage over returning `FooClient` here... And you're getting into C-style deferencing etc business like this line:

    &VirtualMachineImagesClient{*NewWithBaseUri}

this just looks nasty to me. I think we should.

* either do not use pointers at all (my preference)
* (assuming you got a reason to use pointers) change `type VirtualMachineImagesClient struct` to embed `*ComputeManagementClient`; not `ComputeManagementClient` (otherwise you'll do this deferencing here and it just looks ugly.)